private/bufpkg/bufprotoplugin/response_writer.go

```diff
@@ -110,48 +111,43 @@ func writeInsertionPoint(
 	// bufio.Scanner.Scan() inline
 	newline := []byte{'\n'}
 	var found bool
-	for targetScanner.Scan() {
+	for i := 0; targetScanner.Scan(); i++ {
+		if i > 0 {
+			// These writes cannot fail, they will panic if they cannot allocate.
+			_, _ = postInsertionContent.Write(newline)
+		}
+
 		targetLine := targetScanner.Bytes()
 		if !bytes.Contains(targetLine, match) {
-			// these writes cannot fail, they will panic if they cannot
-			// allocate
+			// These writes cannot fail, they will panic if they cannot allocate.
 			_, _ = postInsertionContent.Write(targetLine)
-			_, _ = postInsertionContent.Write(newline)
 			continue
 		}
 		// For each line in then new content, apply the
 		// same amount of whitespace. This is important
 		// for specific languages, e.g. Python.
 		whitespace := leadingWhitespace(targetLine)
 
-		// Create another scanner so that we can seamlessly handle
-		// newlines in a platform-agnostic manner.
-		insertedContentScanner := bufio.NewScanner(bytes.NewBufferString(insertionPointFile.GetContent()))
-		insertedContent := scanWithPrefixAndLineEnding(insertedContentScanner, whitespace, newline)
-		// This write cannot fail, it will panic if it cannot
-		// allocate
-		_, _ = postInsertionContent.Write(insertedContent)
+		// Insert the content from the insertion point file. Handle newlines in
+		// a platform-agnostic manner.
+		insertedContentReader := strings.NewReader(insertionPointFile.GetContent())
+		writeWithPrefixAndLineEnding(postInsertionContent, insertedContentReader, whitespace, newline)
 
 		// Code inserted at this point is placed immediately
 		// above the line containing the insertion point, so
 		// we include it last.
 		// These writes cannot fail, they will panic if they cannot
 		// allocate
 		_, _ = postInsertionContent.Write(targetLine)
-		_, _ = postInsertionContent.Write(newline)
```

Comment:
Do we not need a new line write before the `targetLine`, similar to the non-matching case here?

```
		if i > 0 {
			// These writes cannot fail, they will panic if they cannot allocate.
			_, _ = postInsertionContent.Write(newline)
		}
```

Reply:
I don't think so. Instead of always appending a trailing newline so the next content starts at a newline, we reverse that logic to append a newline if not the first line. This removes the need to strip the newline when we finished writing. Target line will be on a newline as the insert point above has a trailing newline.

Reply:
Ah I see, we did that outside of the conditional, okay cool 👍🏼